add kubevirt SecurityContextConstraints

mfranczy opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. Follow the instructions for writing a release note from k8s: https://git.k8s.io/community/contributors/guide/release-notes.md
-->

**What this PR does / why we need it**:
It handles SecurityContextConstraints objects and creates the `kubevirt` SCC for virt-launcher.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:
Still updating tests.

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
SecurityContextConstraints support
```

mfranczy commented:
@davidvossel @slintes now I am updating tests for virt-operator but this change is reviewable, please share your opinion about the solution introduced here.

davidvossel commented on manifests/generated/virt-api.yaml.in:
```diff
@@ -77,5 +77,5 @@ spec:
         resources: {}
       securityContext:
-        runAsNonRoot: true
+        runAsNonRoot: false
```
Does this have to change?

mfranczy replied:
it doesn't, removed.

davidvossel commented on pkg/virt-operator/application.go:
```diff
@@ -159,10 +163,6 @@ func Execute() {
-	}
 	if onOpenShift {
 		log.Log.Info("we are on openshift")
 		app.informers.SCC = app.informerFactory.OperatorSCC()
```
app.informerFactory.OperatorSCC needs to change to only watch SCC objects with our operator label.

mfranczy replied:
In the current implementation, if I change that then operator won't add virt-handler to privileged SCC (we observe only SCC with operator label, privileged SCC doesn't have required label).

Hm, since I added SCC component, I think it's worth to create an additional `virt-privileged` SCC for virt-handler (and components which need that) and stop modifying the generic `privileged`, it would simplify the implementation (we could drop `addOrRemoveSCCUsers` func) and we would have full control over virt-handler permissions.

In my opinion, it's a good solution, what do you think?

mfranczy replied:
To give you an overview, I pushed that change quickly, https://github.com/kubevirt/kubevirt/pull/2545/files#diff-2c9edb5dc9218f571926c89df6f12435R45 and we got rid of https://github.com/kubevirt/kubevirt/pull/2545/files#diff-9a9763888082c586cc68fce7c54befd2L46

davidvossel replied:
yep, we don't want to modify the `privileged` SCC object anymore.

mfranczy commented:
PR from my perspective is finished.

davidvossel commented on pkg/virt-operator/install-strategy/strategy.go:
```diff
@@ -235,19 +221,10 @@ func GenerateCurrentInstallStrategy(config *operatorutil.KubeVirtDeploymentConfi
-	typeMeta := metav1.TypeMeta{
-		Kind: customSCCPrivilegedAccountsType,
+	// add all OpenShift specific resources to InstallStrategy
+	if config.OnOpenShift {
```
Would it simplify things if we always appended the SCC objects to the strategy.sccs slice, and then only applied them when we're on OpenShift?

mfranczy replied:
yes, I check it once in [createOrUpdateSCC](https://github.com/kubevirt/kubevirt/pull/2545/files#diff-d5561d66dec277d1e552dd8ac98e1cfcR1436)

davidvossel commented on pkg/virt-operator/creation/components/scc.go:
```diff
@@ -0,0 +1,85 @@
+func NewKubeVirtSCC(namespace string) *secv1.SecurityContextConstraints {
+	scc := newBlankSCC()
+
+	scc.Name = "kubevirt"
```
Should this name reflect that the scc rule is just for virt-controller? maybe it should just be called 'kubevirt-controller'

mfranczy replied:
done

davidvossel commented on pkg/virt-operator/creation/components/scc.go:
```diff
@@ -0,0 +1,85 @@
+func NewKubeVirtPrivilegedSCC(namespace string) *secv1.SecurityContextConstraints {
+	scc := newBlankSCC()
+
+	scc.Name = "kubevirt-privileged"
```
maybe name it 'kubevirt-handler' since it only applies to the handler?

mfranczy replied:
done

davidvossel commented on pkg/virt-operator/creation/rbac/operator.go:
```diff
@@ -184,6 +184,8 @@ func NewOperatorClusterRole() *rbacv1.ClusterRole {
 					"list",
 					"watch",
+					"create",
+					"patch",
```
we'll need "delete" here as well.

mfranczy replied:
done

mfranczy commented on pkg/virt-operator/application.go:
```diff
@@ -163,6 +164,9 @@ func Execute() {
 	if err != nil {
 		golog.Fatalf("Error determining cluster type: %v", err)
 	}
+	if err := os.Setenv(util.OnOpenshiftEnvName, strconv.FormatBool(onOpenShift)); err != nil {
```
I am not sure if this is the best way to propagate info that Kubevirt is deployed on OpenShift, with this env I don't have to change config logic too much, but I have mixed feelings about it. Any opinion here?

slintes replied:
mh, I indeed don't like it this way. Can't you just cache the isOnOpenShift value in pkg/virt-operator/util/client.go, where it's discovered?
Other than that David was faster than me ;)

mfranczy replied:
Yeah, I will change it, this solution sucks.

davidvossel replied:
i think we should cache the result somewhere outside of the install strategy.

Caching it in `pkg/virt-operator/util/client.go` would be fine, but that would require synchronization using locking. 

Another option is to add a boolean to the `pkg/virt-operator/util/types.go` files Stores struct called soemthing like "IsOnOpenshift". Then in `pkg/virt-operator/application.go` you could do something like this in the onOpenShift if/else that already exists. 

```
        if onOpenShift {                                                                       
                log.Log.Info("we are on openshift")                           
                app.informers.SCC = app.informerFactory.OperatorSCC()                          
                app.stores.SCCCache = app.informerFactory.OperatorSCC().GetStore()
                app.stores.IsOnOpenshift = true
        } else {
                log.Log.Info("we are on kubernetes")
                app.informers.SCC = app.informerFactory.DummyOperatorSCC()    
                app.stores.SCCCache = app.informerFactory.DummyOperatorSCC().GetStore()
        } 
```

mfranczy replied:
done

davidvossel commented on pkg/virt-operator/install-strategy/create.go:
```diff
@@ -1339,89 +1339,57 @@ func createOrUpdateService(kv *v1.KubeVirt,
-			_, err = scc.SecurityContextConstraints().Patch(sccPriv.TargetSCC, types.JSONPatchType, []byte(fmt.Sprintf("[ %s, %s ]", test, patch)))
+			ops = append(ops, fmt.Sprintf(`{ "op": "replace", "path": "/", "value": %s }`, string(sccBytes)))
+			// replace the SCC by the new one
+			_, err = sec.SecurityContextConstraints().Patch(scc.Name, types.JSONPatchType, generatePatchBytes(ops))
```
i'm not 100% sure this Patch on '/' will work because there's going to be things in the cached SCC's metadata (like creationTImestamp and ResourceVersion) that might cause issues if we attempt to patch them. 

I think it's probably safe to just call an Update() instead of using Patch() for the scc. So this could be as simple as something like this. 

```
scc.ObjectMeta = cachedScc.ObjectMeta.DeepCopy()
injectOperatorMetadata(kv, &scc.ObjectMeta, version, imageRegistry, id)
_, err = sec.SecurityContextConstraints().Update(scc)
```

mfranczy replied:
Changed to update as suggested, the update is safe.

davidvossel commented on pkg/virt-operator/install-strategy/strategy.go:
```diff
@@ -73,7 +59,7 @@ type InstallStrategy struct {
 	deployments []*appsv1.Deployment
 	daemonSets  []*appsv1.DaemonSet
 
-	customSCCPrivileges []*customSCCPrivilegedAccounts
```
unfortunately, I think we have to leave this struct and logic around patching `privileged` scc in for now. Otherwise we break updates.

So the logic that updates the `privileged` SCC needs to stay, but the new logic that generates the InstallStrategy should just not set anything on this array and use the new `sccs` slice instead.

slintes replied:
would a simplified logic be enough? Just removing users from the privileged SCC? We don't have a history of changes here yet, we know what we did to the privileged SCC.

slintes replied:
I think we can even get rid of the struct. loadInstallStrategyFromBytes() will just ignore the old config. And we just remove users from the privileged SCC when creating/updating/deleting the new SCCs. Does that make sense?

davidvossel replied:
how do we know what users to remove without this struct?

davidvossel replied:
also, we need the add logic in the event that we have to restore a degraded install to the previously working version.

slintes replied:
> how do we know what users to remove without this struct?
we just know which users were ever added by our code, so we could hardcode them.
but yes, restoring (or even just installing) N-1 is a good reason why we need to keep the struct and logic for now. Too bad.

mfranczy commented:
/hold

I changed the logic here to keep the both SCC solutions working together, I am testing it now. The SCC informer watches all objects (this must stay like it is) so I recognize the SCCs created by the operator by checking `ManagedByLabel` value.

davidvossel commented:
> I changed the logic here to keep the both SCC solutions working together, I am testing it now. The SCC informer watches all objects (this must stay like it is) so I recognize the SCCs created by the operator by checking ManagedByLabel value.

That change makes sense to me. Sorry this is so complex. 

For safety, I think we should make one small additional change to the virt-operator RBAC.  The "delete", "create", "update" permissions for scc should be limited to the resource names of the new SCC objects being introduced.

The reasoning for this is, it prevents us from even having the ability to accidentally introduce a bug in our code that results in all the OpenShift SCCs from being deleted somehow, which could easily happen given how our operator loops through the SCC cache.

davidvossel commented:
I forget to menting, for the above comment this can be accomplished by filling in the optional "resourceNames" slice on the PolicyRule object assigned to the operator's ClusterRole.

mfranczy commented:
> > I changed the logic here to keep the both SCC solutions working together, I am testing it now. The SCC informer watches all objects (this must stay like it is) so I recognize the SCCs created by the operator by checking ManagedByLabel value.
> 
> That change makes sense to me. Sorry this is so complex.

No worries, that makes thing interesting :)

> For safety, I think we should make one small additional change to the virt-operator RBAC. The "delete", "create", "update" permissions for scc should be limited to the resource names of the new SCC objects being introduced.
> 
> The reasoning for this is, it prevents us from even having the ability to accidentally introduce a bug in our code that results in all the OpenShift SCCs from being deleted somehow, which could easily happen given how our operator loops through the SCC cache.

That's a good point, but `create` cannot be limited to resource name, the change will be applied for `delete` and `update`.

From [k8s doc](https://kubernetes.io/docs/reference/access-authn-authz/rbac/),

> Note that create requests cannot be restricted by resourceName, as the object name is not known at authorization time.

mfranczy commented:
/hold cancel

mfranczy commented:
I think I addressed all issues here, @davidvossel @slintes could you do another review round? Sorry for the rush, I am trying to merge that before code freeze.

davidvossel commented on pkg/virt-operator/creation/components/scc.go:
```diff
@@ -0,0 +1,82 @@
+	scc.SELinuxContext = secv1.SELinuxContextStrategyOptions{
+		Type: secv1.SELinuxStrategyRunAsAny,
+	}
+	scc.AllowedCapabilities = []corev1.Capability{secv1.AllowAllCapabilities}
```
we know the exact capabilities required. From pkg/virt-controller/services/template.go it should be, NET_ADMIN, SYS_NICE, and SYS_RESOURCE.

mfranczy replied:
I had these capabilities before and each functional test failed with `LibvirtError(Code=8, Domain=44, Message='invalid argument: could not find capabilities for arch=x86_64 domaintype=kvm)` ([link to failed tests](https://prow.apps.ovirt.org/view/gcs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/2545/pull-kubevirt-e2e-os-3.11.0/1158790334914760706)) 

I was afraid I will have to add many KVMs capabilities that will make the list very long https://www.kernel.org/doc/Documentation/virtual/kvm/api.txt so I decided to allow all because needed capabilities are defined in template.go so anything beyond that would not be added to the pod and the `kubevirt-controller` SCC is assigned only to `kubevirt-controller` service account.

Although, I will check that again.

davidvossel replied:
umm, I'm not sure what you are encountering there. This AllowedCapabilities list isn't actually impacting what capabilities the Pod receives. It's just allowing virt-controller to create a pod with those capabilities which is something that is already occurring. 

I wouldn't expect the values in this list to impact the VMI at runtime.

mfranczy replied:
you're right, it seems my environment was messed, on fresh cluster it works.

davidvossel commented on pkg/virt-operator/install-strategy/create.go:
```diff
@@ -1425,6 +1426,57 @@ func addOrRemoveSSC(targetStrategy *InstallStrategy,
+			}
+			log.Log.V(2).Infof("SCC %v created", scc.Name)
+		} else if !objectMatchesVersion(&cachedSCC.ObjectMeta, version, imageRegistry, id) {
+			_, err = sec.SecurityContextConstraints().Update(scc)
```
I'm not 100% sure if this Update will always work unless we copy the metadata from the old cachedSCC and then override only the version/registry annotations.

So, something like this. 

```
scc.ObjectMeta = cachedSCC.ObjectMeta.DeepCopy()
injectOperatorMetadata(kv, &scc.ObjectMeta, version, imageRegistry, id)
_, err = sec.SecurityContextConstraints().Update(scc)
if err != nil {
	return fmt.Errorf("Unable to update %s SecurityContextConstraints", scc.Name)
}
log.Log.V(2).Infof("SecurityContextConstraints %s updated", scc.Name)
```

mfranczy replied:
ah yes, you pointed out that already in previous comments, sorry, I will fix it.

mfranczy replied:
done

davidvossel commented on pkg/virt-operator/kubevirt.go:
```diff
@@ -283,6 +296,27 @@ func (c *KubeVirtController) getKubeVirtKey() (string, error) {
+}
+
+func (c *KubeVirtController) sccDeleteHandler(obj interface{}, expecter *controller.UIDTrackingControllerExpectations) {
+	o := obj.(metav1.Object)
```
take a look at the `genericDeleteHandler` function. You'll need to add a similar block there to handle the `cache.DeletedFinalStateUnknown` obj state.

mfranczy replied:
Hm, the thing that `cache.DeletedFinalStateUnknown` is handled in `genericDeleteHandler` is not enough? I am passing the whole SCC object to `genericDeleteHandler` ([link](https://github.com/kubevirt/kubevirt/pull/2545/files#diff-dab38d1bd4e3b7017adcaf4fc06f65e1R316)) if it's managed by the operator, so failed delete process will be reported anyway. Am I missing something here? Or maybe I should check `cache.DeletedFinalStateUnknown` state for all SCC objects (not only created by the operator)?

davidvossel replied:
The problem is with the cast of the obj and the label check that occurs before passing to genericDeleteHandler. We're not guaranteed in the delete handler that the obj is of type metav1.Object, which means the cast could fail and the label check wouldn't cause the genericDeleteHandler function to be invoked.

We have to also check the tombstone object in the delete callback.

mfranczy replied:
ah, that makes perfect sense, looking a second time with a fresh mind made the issue understandable, thanks for the explanation.

mfranczy replied:
done

stu-gott commented:
/lgtm

davidvossel commented:
looks great, We're just waiting on CI now. The `pull-kubevirt-e2e-os-3.11.0` and `pull-kubevirt-e2e-os-3.11.0-crio` lanes are going to be the ones we're most interested in here.

mfranczy commented:
/retest

slintes commented:
test failure is unrelated, great work! :+1:

fabiand commented:
That's crazy work! Well done.
